From 05ab5d62b66de6e131f71de08254dd84ef41f8c2 Mon Sep 17 00:00:00 2001
From: Ken Zhu <quic_guigenz@quicinc.com>
Date: Thu, 4 Apr 2024 16:01:38 -0700
Subject: [PATCH] [net] [bridge] Allow unicast loop back when hairpin enabled

disable multicast go back to the original device for non wireless
device.

Change-Id: Ia5124d2c3d602f71d444f4ac800ddae4b608e4c2
Signed-off-by: Ken Zhu <quic_guigenz@quicinc.com>
---
 net/bridge/br_forward.c | 33 +++++++++++----------------------
 1 file changed, 11 insertions(+), 22 deletions(-)

diff --git a/net/bridge/br_forward.c b/net/bridge/br_forward.c
index b41be2778b3a..0b9a697a42d5 100644
--- a/net/bridge/br_forward.c
+++ b/net/bridge/br_forward.c
@@ -25,28 +25,17 @@ static inline int should_deliver(const struct net_bridge_port *p,
 
 	vg = nbp_vlan_group_rcu(p);
 
-	if ((skb->dev != p->dev) &&
-	    br_allowed_egress(vg, skb) && p->state == BR_STATE_FORWARDING &&
-	    nbp_switchdev_allowed_egress(p, skb) &&
-	    !br_skb_isolated(p, skb)) {
-		return true;
-	}
-
-	if ((skb->dev == p->dev) &&
-	    (p->flags & BR_HAIRPIN_MODE) &&
-	    br_allowed_egress(vg, skb) && p->state == BR_STATE_FORWARDING) {
-
-		/* Allow hairpin only on WLAN netdevices.
-		 * For ethernet interfaces, hairpin not allowed
-		 * due to issue with switchdevs. */
-		if ( skb->dev->ieee80211_ptr == NULL) {
-			return false;
-		}
-
-		return true;
-	}
-
-	return false;
+	/*
+	 * When hairpin enabled, don't allow  multicast go back
+	 * to the original non-wireless device.
+	 * it could cause fdb learning issue to connected switch.
+	 */
+	return (((p->flags & BR_HAIRPIN_MODE)
+		&& (skb->dev->ieee80211_ptr != NULL || !is_multicast_ether_addr(eth_hdr(skb)->h_dest)))
+		|| (skb->dev != p->dev)) &&
+		br_allowed_egress(vg, skb) && (p->state == BR_STATE_FORWARDING) &&
+		nbp_switchdev_allowed_egress(p, skb) &&
+		!br_skb_isolated(p, skb);
 }
 
 int br_dev_queue_push_xmit(struct net *net, struct sock *sk, struct sk_buff *skb)
-- 
2.34.1

